### PR TITLE
Fix gevent LoopExit exception when all greenlets from a service exit.

### PR DESCRIPTION
**Pull request by danielrowles-wf**

### CODE REVIEW

File: biloba/test/test_service.py [hunk 1]
```diff
@@ -10,6 +10,33 @@
...
+        """
+        self.spawn(self.noddy_likes_pies)
+
+    def noddy_likes_pies(self):
+        """
+        Run noddy service
+        """
+        while True:
+            for s in ('noddy', 'likes', 'pies'):
+                print "TEST: %s" % s
```
Comment (nickjoyce-wf) [inline biloba/test/test_service.py]: Remove print statement

**Comment by kevinfrommelt-wf** [comment]
If the service doesn't spawn anything in the do_start, will the service just start and stop immediately?

**Comment by danielrowles-wf** [comment]
Yes, although I think that's probably what should happen in this case, and that the current behaviour is wrong :) This is going to need careful testing though as a result.

**Comment by kevinfrommelt-wf** [comment]
Should we add a test(s) that for starting a service that doesn't spawn anything or have any child services?

**Comment by kevinfrommelt-wf** [comment]
`Service.join` needs to be updated to `join` on the `pool` rather than `_run_thread`

**Comment by danielrowles-wf** [comment]
I'm not sure it does - I think service.join shouldn't return until after the cleanup of the service has been completed...

File: biloba/service.py [hunk 2]
```diff
@@ -52,7 +52,6 @@ def __init__(self, logger=None):
         self.pool = pool.Group()
         self.logger = logger or self.get_logger()
         self._run_thread = None
-        self._kill = event.Event()
```
Comment (njoyce) [inline biloba/service.py]: Remove the ``_kill`` slot too.

File: biloba/test/test_service.py (hunk 1 above)
Comment (nickjoyce-wf) [inline biloba/test/test_service.py]: use ``.format()`` please

**Comment by nickjoyce-wf** [comment]
service.join should not return until the service and all of it's child services are completed.

**Comment by njoyce** [comment]
``service.join`` should not return until the service and all of it's child services are completed.

File: biloba/service.py [hunk 3]
```diff
@@ -143,7 +139,7 @@ def stop(self, block=True):
         if not self.started:
             return
 
-        self._kill.set()
+        self.pool.kill(block=block)
```
Comment (njoyce) [inline biloba/service.py]: I don't think this is the point at which to block when calling stop - let ``_run_thread.get`` do that.

Comment (njoyce) [inline biloba/service.py]: In fact according to https://github.com/gevent/gevent/blob/master/gevent/pool.py#L101 - block is ``True`` by default - this should explictly be called as ``block=False`` here.

**Comment by njoyce** [comment]
This is superceded by #8

**Pull request closed without merging**